### collectMetrics.R ###############################################################################
# Summarize fragments (number and median size) for REF and ALT reads and compare fragment size
# distribution using a KS test.

### collectMetrics #################################################################################
#' Summarize fragment size
#'
#' This function returns summary metrics, including counts and median fragment size, and attempts 
#' to classify the target variant as somatic, germline or non-somatic.
#'
#' @param fragment.data list containing annotated fragment data generated by annotateReads
#' @param target data.frame containing target mutation (see formatTargets)
#' @param verbose logical indicating whether or not to print metrics to screen (default = TRUE)
#' @return data.frame of summary metrics 
#' @examples
#' test.maf.file <- system.file('extdata', 'test_snp_data.maf', package = 'CompareFragmentSize'); 
#' test.bam.file <- system.file('extdata', 'sample1_example.bam', package = 'CompareFragmentSize'); 
#' mutations <- formatTargets(input = test.maf.file, id = 'Sample1');
#' target.reads <- getReadsFromBAM(filePath = test.bam.file, targets = mutations[1,]);
#' fragment.data <- annotateReads(ga = target.reads, targets = mutations[1,]);
#' summary.metrics <- collectMetrics(fragment.data, mutations[1,]);
#' @export
collectMetrics <- function(fragment.data, target, verbose = TRUE) {

	if ('list' != class(fragment.data)) {
		stop('fragment.data should be a list output by annotateReads.');
	} else if (!'annotated' %in% names(fragment.data)) {
		stop('fragment.data should be a list output by annotateReads.');
	} else {
		fragments <- fragment.data$annotated;
	}

	# count and summarize insert size per allele 
	variant <- paste(target[1,c('Chromosome','Start','REF','ALT')], collapse = '_');

	ref.fragments <- fragments[which(fragments$Group == 'REF')];
	alt.fragments <- fragments[which(fragments$Group == 'ALT')];

	if (verbose) {
		print(paste('For variant: ', variant));
		print(paste('N reference reads: ', length(ref.fragments)));
		print(paste('N variant reads: ', length(alt.fragments)));
		print(aggregate(isize ~ Allele, fragments, median));
		}

	output.data <- target[1,];
	output.data[,c('N.Ref','Median.Ref','N.Alt','Median.Alt','p','VAF','Classification')] <- NA;

	# summarize fragment sizes
	if (length(ref.fragments) > 0) {
		output.data[1,]$N.Ref <- length(ref.fragments);
		output.data[1,]$Median.Ref <- median(ref.fragments$isize);
		}

	if (length(alt.fragments) > 0) {
		output.data[1,]$N.Alt <- length(alt.fragments);
		output.data[1,]$Median.Alt <- median(alt.fragments$isize);
		}

	if ( (length(ref.fragments) > 0) & (length(alt.fragments) > 0) ){
		output.data[1,]$p <- ks.test(
			alt.fragments$isize,
			ref.fragments$isize,
			alternative = 'g')$p.value;

		output.data[1,]$VAF <- output.data[1,]$N.Alt / sum(output.data[1,c('N.Ref','N.Alt')]);

		# indicate predicted classification for each variant (somatic [tumour-derived] or 
		#	non-somatic [CHIP/germline/sequencing artefact])
		output.data$Classification <- if (output.data[1,]$p < 0.01) { 'somatic';
			} else if (output.data[1,]$VAF >= 0.38) { 'probable germline';
			} else { 'non-somatic';
			}

		if (verbose) {
			print(paste('Classification:', output.data$Classification[1]));
			}
		}

	return(output.data);
	}
