### plotFragmentSizes.R ############################################################################
# Create cumulative fragment size distribution plots.

### plotFragmentSizes ##############################################################################
#' Plot cumulative fragment size distributions
#'
#' This function generates a cumulative distribution plot comparing fragment size between
#' REF- and ALT-containing reads using ggplot2.
#'
#' @param fragment.data list containing annotated fragment data generated by annotateReads
#' @param id sample ID; used for plot title (default = NULL)
#' @param target data.frame containing target mutation (see formatTargets)
#' @param output.to should the plot be output to file (as a png) or output to screen (default = 'file')
#' @return png file or print image to screen
#' @examples
#' test.maf.file <- system.file('extdata', 'test_snp_data.maf', package = 'CompareFragmentSize'); 
#' test.bam.file <- system.file('extdata', 'sample1_example.bam', package = 'CompareFragmentSize'); 
#' mutations <- formatTargets(input = test.maf.file, id = 'Sample1');
#' target.reads <- getReadsFromBAM(filePath = test.bam.file, targets = mutations[1,]);
#' fragment.data <- annotateReads(ga = target.reads, targets = mutations[1,]);
#' summary.metrics <- collectMetrics(fragment.data, mutations[1,]);
#' #not run: plotFragmentSizes(fragment.data, 'Sample1', summary.metrics);
#' @export
plotFragmentSizes <- function(fragment.data, target, id = NULL, output.to = 'file') {

	# check fragment data
	if ('list' != class(fragment.data)) {
		stop('fragment.data should be a list output by annotateReads.');
	} else if (!'annotated' %in% names(fragment.data)) {
		stop('fragment.data should be a list output by annotateReads.');
	} else {
		fragments <- fragment.data$annotated;
	}

	# ensure proper output.to
	if (!output.to %in% c('file','screen')) {
		stop("output.to must be one of 'file' or 'screen'.");
		}

	# set theme
	theme <- ggplot2::theme(
		plot.title = ggplot2::element_text(hjust = 0.5, size = 15, face = "bold", family = 'Arial'),
		axis.line = ggplot2::element_line(colour = "black"),
		panel.grid.major = ggplot2::element_blank(),
		panel.grid.minor = ggplot2::element_blank(),
		panel.background = ggplot2::element_blank(),
		legend.position = "none",
		legend.key = ggplot2::element_rect(fill = "white"),
		axis.text = ggplot2::element_text(size = 10, family = 'Arial'),
		axis.title = ggplot2::element_text(size = 12, family = 'Arial'),
		axis.text.x = ggplot2::element_text(angle = 0, colour = 'black', family = 'Arial'),
		axis.text.y = ggplot2::element_text(angle = 0, colour = 'black', family = 'Arial')
		);

	# check ggplot2 version to ensure proper arguments are used
	if (-1 == compareVersion(packageDescription('ggplot2')$Version,'3.4.0')) {
		theme$panel.border <- ggplot2::element_rect(colour = "black", fill = NA, size = 0.5);
		} else {
		theme$panel.border <- ggplot2::element_rect(colour = "black", fill = NA, linewidth = 0.5)
		}

	# use target information to create plot title and output file name
	variant <- paste(target[1,c('Chromosome','Start','REF','ALT')], collapse = '_');

	title <- if ( ('Hugo_Symbol' %in% colnames(target)) & ('HGVSc' %in% colnames(target)) ) {
		paste0(id, ' (', target[1,]$Hugo_Symbol, ': ', target[1,]$HGVSc, ')');
		} else {
		paste0(id, ' (', variant, ')');
		}

	if (nchar(title) >= 40) { title <- paste0(id, ' (', variant, ')'); }

	# collect fragment size metrics
	ref <- data.frame(V1 = fragments[which(fragments$Group == 'REF')]$isize);
	alt <- data.frame(V1 = fragments[which(fragments$Group == 'ALT')]$isize);
	wt <- length(ref$V1);
	mut <- length(alt$V1);
	median1 <- median(ref$V1, na.rm = TRUE);
	median2 <- median(alt$V1, na.rm = TRUE);

	vaf <- round(target[1,]$VAF,2);
	p <- format.pvalue(target[1,]$KS.p);

	# Plot cumulative distributions
	if (-1 == compareVersion(packageDescription('ggplot2')$Version,'3.4.0')) {
		cfd.plot <- ggplot2::ggplot() + 
			ggplot2::stat_ecdf(data = ref, ggplot2::aes(V1, color = "black"), geom = "step") +
			ggplot2::stat_ecdf(data = alt, ggplot2::aes(V1, color = "red"), geom = "step") +
			ggplot2::xlab("Fragment Length (bases)") + 
			ggplot2::ylab("Cumulative Distribution") +
			ggplot2::ggtitle(title) + 
			ggplot2::scale_colour_manual(name = 'Fragments',
				values = c("black" = "black", "red" = "red"), labels = c("Wildtype", "ALT")) +
			theme + 
			ggplot2::geom_vline(xintercept = median1, linetype = "dashed",
				color = "black", size = 0.5) +
			ggplot2::geom_vline(xintercept = median2, linetype = "dashed",
				color = "red", size = 0.5) +
			ggplot2::annotate("text", x = 220, y = c(0.27, 0.20, 0.13, 0.06), 
				label = c(p, paste0("WT = ", wt), paste0("Mutant = ", mut),
					paste0("VAF = ", vaf)),
				col = c("black", "black", "red", "black"),
				size = 4);

		} else {

		cfd.plot <- ggplot2::ggplot() + 
			ggplot2::stat_ecdf(data = ref, ggplot2::aes(V1, color = "black"), geom = "step") +
			ggplot2::stat_ecdf(data = alt, ggplot2::aes(V1, color = "red"), geom = "step") +
			ggplot2::xlab("Fragment Length (bases)") + 
			ggplot2::ylab("Cumulative Distribution") +
			ggplot2::ggtitle(title) + 
			ggplot2::scale_colour_manual(name = 'Fragments',
				values = c("black" = "black", "red" = "red"),
				labels = c("Wildtype", "ALT")) +
			theme + 
			ggplot2::geom_vline(xintercept = median1, linetype = "dashed",
				color = "black", linewidth = 0.5) +
			ggplot2::geom_vline(xintercept = median2, linetype = "dashed",
				color = "red", linewidth = 0.5) +
			ggplot2::annotate("text", x = 220, y = c(0.27, 0.20, 0.13, 0.06), 
				label = c(p, paste0("WT = ", wt), paste0("Mutant = ", mut),
					paste0("VAF = ", vaf)),
				col = c("black", "black", "red", "black"),
				size = 4);
		}

	# should this plot be saved to file or shown on screen?
	if ('file' == output.to) {

		outfile <- paste0('fragment_sizes_', variant, '.png');
		ggplot2::ggsave(
			outfile,
			cfd.plot,
			units = 'in',
			width = 4.5,
			height = 3,
			dpi = 300
			);

		} else { cfd.plot; }
	}
